Sources/System/AsyncFileDescriptor.swift

```diff
@@ -0,0 +1,166 @@
+@_implementationOnly import CSystem
+
+public struct AsyncFileDescriptor: ~Copyable {
+    @usableFromInline var open: Bool = true
+    @usableFromInline let fileSlot: IORingFileSlot
+    @usableFromInline let ring: ManagedIORing
+
+    public static func openat(
+        atDirectory: FileDescriptor = FileDescriptor(rawValue: -100),
+        path: FilePath,
+        _ mode: FileDescriptor.AccessMode,
+        options: FileDescriptor.OpenOptions = FileDescriptor.OpenOptions(),
+        permissions: FilePermissions? = nil,
+        onRing ring: ManagedIORing
+    ) async throws -> AsyncFileDescriptor {
+        // todo; real error type
+        guard let fileSlot = ring.getFileSlot() else {
+            throw IORingError.missingRequiredFeatures
+        }
+        let cstr = path.withCString {
+            return $0  // bad
+        }
+        let res = await ring.submitAndWait(
+            .openat(
+                atDirectory: atDirectory,
+                path: cstr,
+                mode,
+                options: options,
+                permissions: permissions,
+                intoSlot: fileSlot.borrow()
+            ))
+        if res.result < 0 {
+            throw Errno(rawValue: -res.result)
+        }
+
+        return AsyncFileDescriptor(
+            fileSlot, ring: ring
+        )
+    }
+
+    internal init(_ fileSlot: consuming IORingFileSlot, ring: ManagedIORing) {
+        self.fileSlot = consume fileSlot
+        self.ring = ring
+    }
+
+    @inlinable @inline(__always)
+    public consuming func close(isolation actor: isolated (any Actor)? = #isolation) async throws {
```

Comment:
I want to review the heavy use of isolation parameters at some point, as well as inlinability